Pin macOS version on CI

nilsvu opened the pull request:
Recently GitHub moved to macOS 12, which made all of our unit tests time out for some reason (maybe the switch to AppleClang 13 -> 14, though I'm building locally with AppleClang 14 just fine). Pin to macOS 11 until we can fix this.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Pin macOS version on CI

Recently GitHub moved to macOS 12, which made all of our unit tests
time out for some reason (maybe the switch to AppleClang 13 -> 14).
Pin to macOS 11 until we can fix this.

knelli2 reviewed (dismissed):
I was noticing this as well, thanks for doing this

nilsvu committed:
Constrain version of 'rich'

Choose rich >= 10.11.
Rich v12.0 isn't in Spack v0.18 yet, and upgrading to
Spack v0.19 isn't trivial.

nilsvu commented:
@knelli2 for some reason the BoundaryMessage test reports "subprocess aborted". Any idea why that might be or what's special about the BoundaryMessage test?

nilsvu commented:
The error message is empty for some reason and the test looks like it successfully ran, but it still reports "subprocess aborted" and fails. Searching for this I found a case where the failure was "malloc: pointer being freed was not allocated". That could be related maybe?

knelli2 commented:
@nilsvu So the BoundaryMessage class does some pointer math and `reinterpret_cast`ing internally. I wonder if a pointer or something isn't being cleaned up properly in the test? I don't have access to a Mac so I can't test for myself, but am happy to debug this with you

nilsvu commented:
I opened #4610 for the BoundaryMessage test failure. I suggest we merge this PR now to revive CI and Kyle and I work on fixing BoundaryMessage in a follow up.

knelli2 reviewed (approved):
If we merge this, Mac CI will run again, but will still fail on the boundary message test. I guess that's better than no tests being run on Mac.

nilsvu commented:
Yea, currently the macOS CI test runs for 6 hours and times out, that's a lot of wasted CI time. Better have it fail the BoundaryMessage test until we fix it.
